### DISCUSSION TITLE
Feedback: testing multiple modules; leftover app.storage artifacts after tests

### DISCUSSION THREAD

**Discussion by nastasie-octavian** [zauberzeug/nicegui#3649] [Ideas / Feature Requests]
I've been trying out the user fixture over the past few days to add tests to my NiceGUI app and I have some feedback that I'd like to share.

1. There should be a way/fixture to test multiple modules at the same time

    For projects that have multiple modules (e.g. a login module with the login page and other authentication related functions and endpoints, and one or more other modules with other pages and functions), if we want to test the interaction between pages in different modules (e.g. login pages redirects the user to some page in another module after the user logs in, or some page redirects the user to the login page which is in a different module) the `module_under_test()` marker is not enough and the test functions need to have additional `importlib.reload()` calls for the other modules.

    I think that it makes sense to add a `modules_under_test()` marker that can take and reload an arbitrary number of modules for testing.

3. Tests should remove the .nicegui directory created during testing after they finish

    When a page that is being tested stores some data using `app.storage` and a .nicegui directory with a .json file is created, the directory and the file remain on the filesystem after the test is over. I think that the testing framework should ensure that these are deleted after a test is over. 

   It's a minor issue since `app.reset()` (and hence `app.storage.clear()`) is called before each test so you don't end up with 1000s of .json files, but there's something about leaving transient testing artifacts on disk that's bugging me.

**Comment by rodja** [comment]
Thanks for your feedback @nastasie-octavian. We are always happy to hear about experiences with the new `user` fixture.

> I think that it makes sense to add a modules_under_test() marker that can take and reload an arbitrary number of modules for testing.

The `module_under_test` marker was mainly introduced to simplify testing of small apps which fit in a single file. If you have a larger app with multiple modules I suggest you have a look at https://nicegui.io/documentation/project_structure#modular which describes a more modular approach. The main benefit is to reduce boilerplate code for every test and more control over what should be loaded in the first place.

> Tests should remove the .nicegui directory created during testing after they finish

You are right. I just created #3663.
